solr/contrib/gcs-repository/src/java/org/apache/solr/gcs/GCSBackupRepository.java

```diff
@@ -0,0 +1,466 @@
+             IndexOutput output = dest.createOutput(destFileName, DirectoryFactory.IOCONTEXT_NO_CACHE)) {
+            ByteBuffer buffer = ByteBuffer.allocate(1024 * 1024 * 8);
+            while (readChannel.read(buffer) > 0) {
+                buffer.flip();
+                byte[] arr = buffer.array();
+                output.writeBytes(arr, buffer.position(), buffer.limit() - buffer.position());
+                buffer.clear();
+            }
+        }
+
+    }
+
+
+    @Override
+    public void close() throws IOException {
+
+    }
+
+    private void writeBlobMultipart(BlobInfo blobInfo, ChecksumIndexInput indexInput, int blobSize)
+            throws IOException {
+        byte[] bytes = new byte[blobSize];
+        indexInput.readBytes(bytes, 0, blobSize - CodecUtil.footerLength());
+        long checksum = CodecUtil.checkFooter(indexInput);
+        ByteBuffer footerBuffer = ByteBuffer.wrap(bytes, blobSize - CodecUtil.footerLength(), CodecUtil.footerLength());
+        writeFooter(checksum, footerBuffer);
+        try {
+            storage.create(blobInfo, bytes, Storage.BlobTargetOption.doesNotExist());
+        } catch (final StorageException se) {
+            if (se.getCode() == HTTP_PRECON_FAILED) {
+                throw new FileAlreadyExistsException(blobInfo.getBlobId().getName(), null, se.getMessage());
+            }
+            throw se;
+        }
+    }
+
+    private void writeBlobResumable(BlobInfo blobInfo, ChecksumIndexInput indexInput) throws IOException {
+        try {
+            final Storage.BlobWriteOption[] writeOptions = new Storage.BlobWriteOption[0];
+            final WriteChannel writeChannel = storage.writer(blobInfo, writeOptions);
+
+            ByteBuffer buffer = ByteBuffer.allocate(BUFFER_SIZE);
+            writeChannel.setChunkSize(BUFFER_SIZE);
+
+            long remain = indexInput.length() - CodecUtil.footerLength();
+            while (remain > 0) {
+                // reading
+                int byteReads = (int) Math.min(buffer.capacity(), remain);
+                indexInput.readBytes(buffer.array(), 0, byteReads);
+                buffer.position(byteReads);
+                buffer.flip();
+
+                // writing
+                writeChannel.write(buffer);
+                buffer.clear();
+                remain -= byteReads;
+            }
+            long checksum = CodecUtil.checkFooter(indexInput);
+            ByteBuffer bytes = getFooter(checksum);
+            writeChannel.write(bytes);
+            writeChannel.close();
+        } catch (final StorageException se) {
+            if (se.getCode() == HTTP_PRECON_FAILED) {
+                throw new FileAlreadyExistsException(blobInfo.getBlobId().getName(), null, se.getMessage());
+            }
+            throw se;
+        }
+    }
+
+    private ByteBuffer getFooter(long checksum) throws IOException {
+        ByteBuffer buffer = ByteBuffer.allocate(CodecUtil.footerLength());
+        writeFooter(checksum, buffer);
```

Comment:
These two methods could be grouped.

Reply:
Not quite sure your meaning here.  Are you pointing out that buffer allocation could occur in `writeFooter` since that method returns the buffer?

If I'm reading your point correctly, I'm not sure I agree. There's at least one call to `writeFooter` that's able to use an existing bytes[] and provide it's own buffer.  It wouldn't be the end of the world to scrap that, but I think keeping buffer allocation and filling separate makes some sense at least.

Anyway, lmk what you think.

Reply:
Ok I missed the direct call to writeFooter, so never mind my previous comment.